### PR TITLE
core/pom.xml: Update Groovy to v2.5.x

### PR DESCRIPTION
**Pull request by allanlewis**
See [JENKINS-53372](https://issues.jenkins-ci.org/browse/JENKINS-53372).

### Proposed changelog entries

* Upgrade bundled Groovy to 2.5.2.

### Submitter checklist

- [/] JIRA issue is well described
- [/] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [/] Appropriate autotests or explanation to why this change has no tests
- [/] For dependency updates: links to external changelogs and, if possible, full diffs
      * See http://groovy-lang.org/changelogs.html

### CODE REVIEW

**Comment by daniel-beck** [comment]
@daspilker WDYT?

**Comment by jglick** [comment]
At a minimum requires prior running of all `workflow-cps` and `workflow-cps-global-lib` tests against the patched core.

**Comment by jglick** [comment]
Test failures like

```
java.lang.NoSuchMethodError: org.codehaus.groovy.runtime.memoize.EvictableCache.getAndPut(Ljava/lang/Object;Lorg/codehaus/groovy/runtime/memoize/MemoizeCache$ValueProvider;)Ljava/lang/Object;
	at org.codehaus.groovy.macro.transform.MacroMethodsCache.get(MacroMethodsCache.java:50)
```

suggest something amiss in the POM, perhaps.

**Review by svanoort — CHANGES_REQUESTED** [review]
At risk of being the bad guy here: why?  Every bump to Groovy is a very high-risk change and often introduces critical problems. For example, the bump to 2.4.12 to support Java 10 introduced a nasty memory leak with Pipelines (still not fully resolved -- time consuming). 

I would not suggest a change like this without a good reason and  *very extensive* testing.

**Comment by allanlewis** [comment]
> At risk of being the bad guy here: why?

Not at all, you're being perfectly reasonable! I just wanted to raise this because I couldn't see it being proposed anywhere in JIRA and I wanted to understand what the blockers might be to this upgrade. If this needs lots of testing, please point me in the right direction and I'll see what I can do. If maintainers think this is too high risk for the limited benefits, I'm happy to close the PR and leave the comments as a record of why it was abandoned.

**Comment by oleg-nenashev** [comment]
The ticket referenced by @svanoort is https://issues.jenkins-ci.org/browse/JENKINS-52023 
I agree with Sam that we need to be careful with the upgrade, but we need to upgrade eventually anyway. So far it mostly depends on the memory leaks as @svanoort pointed out

I would propose to keep the PR open for now

**Comment by svanoort** [comment]
+1 - I'm all for keeping it open.  I think once we have the known issue from 2.4.12 resolved, we can look at when the next Groovy bump will be and what to bump it to.

**Comment by oleg-nenashev** [comment]
One update from the Java 10+ hackathon: https://github.com/jenkinsci/jenkins/pull/3643

**Comment by oleg-nenashev** [comment]
It's no longer on hold. I have resolved the merge conflict so that we see whether it fails now

**Comment by svanoort** [comment]
I mean, we'd have to do a fair bit of testing to see what this does or doesn't break.  2.4.12 seems to have settled out by other means (crossing fingers we don't make any late discoveries here). 

Would still like to see the reasons to move to 2.5.x though -- [2.5.x doesn't seem to have anything pertinent to Jenkins though, and does have a fairly long list of breaking changes](http://groovy-lang.org/releasenotes/groovy-2.5.html).  None of the breaking changes jump out as obvious issues but still that's extra risk.

Now, I could see an argument for Groovy 3.x being worth the testing and risk due to the [new features, including a newer, cleaner (and probably faster) parser and Java 8 syntax support](http://groovy-lang.org/releasenotes/groovy-3.0.html).  It's still in testing but due soonish-ly.   I'd argue strongly for that one being worth the risks.

**Review by svanoort — CHANGES_REQUESTED** [review]
See comment just-now:  the argument for a bump to 2.5.x isn't very strong AFAICT, and would suggest we should wait until 3.x to bump the Groovy version because that has a lot of goodies that might benefit use in Jenkins.

**Comment by allanlewis** [comment]
I've rebased this on master to clear the conflicts. I've also updated Groovy to the latest 2.5.x in a squash commit; this hasn't been tested.

File: core/pom.xml
Comment (oleg-nenashev) [inline core/pom.xml]: This is a bad idea. We include JARs here

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
It won't be passing when the dependency type is POM. 

If you need exclusions here, most likely the required version of ant is higher than the one currently included into the Jenkins WAR. In such case it will cause unpredictable behavior of Groovy. It is better to update dependencies to prevent conflicts than to add exclusions

**Comment by daniel-beck** [comment]
> On-holding this until after the security update next week.

Oops. Fixed.

**Comment by daniel-beck** [comment]
Before we can merge this, we need a dedicated testing effort especially around pipeline, with pipeline load tests (memory leaks etc.). Additionally, once that's done, I strongly recommend the Jenkins security team spend some time trying to break this (looking at you, Script Security).

Anyone here up to take on parts of this?

**Review by jglick — COMMENTED** [review]
I do not think we want to do this. The risk to `script-security` / `workflow-cps` outweighs the benefit of any features.

There is a longstanding RFE to remove Groovy from core altogether, and create a library plugin with a freely floating version that could be used from trusted (`RUN_SCRIPTS`) code. The plugins mentioned above would want to pin their own version.

**Comment by jvz** [comment]
My main concern is how do macros relate to AST transformations since that caused issues in the past.

**Comment by daniel-beck** [comment]
I'm closing this, as it is not practical at the moment. This open PR serves no purpose, as there's simply no way we'd be able to merge a single PR to core with a change like this.

As previous comments in this thread explained, updating Groovy in Jenkins is a major undertaking, affecting a lot of different components. Such a change is a project with many stakeholders in itself, and should probably be discussed on the developers' list and potentially as a JEP first, as it needs to be planned carefully.

File: core/pom.xml
Comment (digulla) [reply core/pom.xml:439]: I agree. Using the POM / BOM approach would be a good idea if this was inside of `dependencyManagement` but it isn't.

**Pull request closed without merging**
